### PR TITLE
represent numerical values in CBMC trace in hex

### PR DESCRIPTION
**Pull request by polgreen**
This adds an option to represent numeric values in the CBMC trace in hex

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: f1cfe4402f0ba1d25a410c53baa7789dea9398e6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77958102

**Comment by polgreen** [comment]
The error on this appears to come because CBMC on Appveyor represents an unsigned int as an unsigned long, e.g., instead of `0u` it outputs `0ul`. That seems incorrect?

**Comment by smowton** [comment]
I'm guessing this is because on all Microsoft environments `unsigned int` and `unsigned long` are the same size (https://msdn.microsoft.com/en-us/library/s3f49ktz.aspx)

**Comment by polgreen** [comment]
Sure, I realise they are the same size, but it still seems like a bug to output long for everything. If nothing else it adds noise when trying to follow through an error trace.

**Comment by smowton** [comment]
Hmmm, looks like the c-type for constants is guessed from their width, here: https://github.com/diffblue/cbmc/blob/6566d1094290ce13a5e25b3fa2710d2b7eb9fcd1/src/ansi-c/literals/convert_integer_literal.cpp#L86

Obviously by guessing it can't give the right answer when `int` and `long int` have the same width. To get it right would need us to infer which one was meant from interactions with explicit variables-- in the particular case of a trace we could force the c-type of an RHS constant to match that of an LHS variable?

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 94781045970a9e52764e9b73b9d0571d8da2e49d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77990353

**Comment by polgreen** [comment]
I've tweaked the regex now so that it works on windows and on linux

**Comment by smowton** [comment]
Yeah, I couldn't be arsed wading into this either :)

File: src/goto-programs/goto_trace.cpp [hunk 1]
```diff
@@ -104,9 +104,42 @@ void goto_trace_stept::output(
...
+    if(++i % 8 == 0 && result.size() != i)
+      oss << ' ';
+  }
+  if(options.base_prefix)
+    return prefix + oss.str();
+  else
+    return oss.str();
+}
+
+std::string trace_value_lowlevel(
```
Comment (smowton) [inline src/goto-programs/goto_trace.cpp]: Suggest just `trace_value`, as `lowlevel` suggests this is a fallback like `expr2lisp` or something

File: src/goto-programs/goto_trace.cpp [hunk 2]
```diff
@@ -104,9 +104,42 @@ void goto_trace_stept::output(
...
   const exprt &expr,
```
Comment (smowton) [inline src/goto-programs/goto_trace.cpp]: This could use a Doxy comment explaining what format you'll get and the customisations available via `options`

**Review by smowton — APPROVED** [review]
LGTM, requesting documentation of one function

File: src/goto-programs/goto_trace.cpp (hunk 1 above)
Comment (polgreen) [reply src/goto-programs/goto_trace.cpp]: There's already a trace_value function so I think introducing another one would make this pretty confusing. Is `trace_numeric_value` any better?

Comment (smowton) [reply src/goto-programs/goto_trace.cpp]: Yeah that's fine, or `integer_value` if this is int types only

**Comment by kroening** [comment]
@smowton The width guessing in the parser is correct: the ISO C standard says so.

The reason for the problem with the type of the literals is more subtle, unfortunately. On windows, the width of long and int are the same; the types only differ in a comment #c_type.  The following happens: After symbolic simulation, we hash the expressions, using a hash that ignores the comments. Thus, on Windows, they get hashed to the same irep. I think we need to promote that comment to be a proper irep field.

**Review by kroening — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 9c1eb1a47738a9dc19fe352884f1d56a9344e4b1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78078212

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: a0186528858b521d32450f774875e53e03c05863).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78157896

**Pull request merged**